### PR TITLE
Upgrade Speech to V1

### LINKED ISSUE #1205: Document Speech low-level API (GAPIC code)
Types such as [v1beta1/speech_api.rb](https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-speech/lib/google/cloud/speech/v1beta1/speech_api.rb) are not currently published to the gh-pages documentation.

- [] Remove exclusions from `.yardopts`
- [] Add navigation to GAPIC/gRPC code

### LINKED ISSUE #1207: Cloud Speech Ruby Veneer:  Avoid "Job"
The veneer should avoid the use of the term “job” because this could cause confusion with first-class Job potentially added in a future version of the API. The V1 API will refer to this as LongRunningRecognize

### LINKED ISSUE #1209: Cloud Speech Ruby Veneer: Sample code doesn't show how to wait until recognition is complete
In stream.rb, none of the samples show how to wait until recognition completes; in fact they will potentially exit early before all processing is done.  Could you illustrate best practices in Ruby for doing this?

### LINKED ISSUE #1214: Add Long Running Operations veneer
Since asynchronous Speech recognize relies on Long Running Operations to communicate when processing is complete, it would be very handy to have a veneer around long-running ops as well.

```wait_until_done!``` is probably not going to get a lot of production use since that would tie up threads for a very long time, indeed.

Thanks for the excellent library @quartzmo

### PR DESCRIPTION
**Pull request by blowmage**
This PR upgrades Speech to use the V1 API. There are several noticeable changes from V1beta1:

* The `language` argument is now mandatory on all recognition methods
* `#recognize_job` has been renamed `#process`, aliased as `#long_running_recognize` and `#recognize_job`
* `#stream` has been aliased as `#stream_recognize`
* `Stream` added `#wait_until_complete!` method, which blocks until stream is complete
* `Stream` removed `#speech_start` and `#speech_end` callback methods
* `Stream` now calls `#complete` callback method when stream ends, not when `END_OF_AUDIO` event is received (the `END_OF_AUDIO` event was removed in V1)
* `Job` has been renamed `Operation`
* `Operation` added `id` attribute, can be retrieved by `id`

This PR closes the following open issues:

* Rename `Job` to `Operation` (closes #1207)
* Update code examples to show how to wait until recognition is complete (closes #1209)
* Add Long Running Operations veneer (closes #1214)
* Document Low level API (GAPIC) (closes #1205)

### CODE REVIEW

**Comment by blowmage** [comment]
@pbogle We would love to get your input on this as well!

**Comment by pbogle** [comment]
There are other changes and improvements as well.  Adding a link is probably best rather than trying to duplicate the release notes.  Please coordinate via email to figure out how we want to do that linking.

File: google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json [hunk 1]
```diff
@@ -1,40 +1,40 @@
...
+    "google.cloud.speech.v1.Speech": {
       "retry_codes": {
-        "idempotent": [
-          "DEADLINE_EXCEEDED",
-          "UNAVAILABLE"
-        ],
-        "non_idempotent": [
-          "UNAVAILABLE"
-        ]
+          "idempotent": [
```
Comment (pbogle) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: Can you point me at documentation for how the idempotent and non_idempotent configuration works?  

I was wondering why UNAVAILABLE appears on both idempotent and non_idempotent list. There are other potential error codes where retries would be appropriate, e.g. INTERNAL.  The API is pretty much stateless, so retrying is always safe.  (Retrying calls might result in duplicated work but it won't mutate state twice or out of order.)

File: google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json [hunk 2]
```diff
@@ -1,40 +1,40 @@
...
         "default": {
           "initial_retry_delay_millis": 100,
           "retry_delay_multiplier": 1.3,
           "max_retry_delay_millis": 60000,
-          "initial_rpc_timeout_millis": 60000,
+          "initial_rpc_timeout_millis": 20000,
           "rpc_timeout_multiplier": 1.0,
-          "max_rpc_timeout_millis": 60000,
+          "max_rpc_timeout_millis": 20000,
           "total_timeout_millis": 600000
```
Comment (pbogle) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: What does it mean to make this 20,000 milliseconds?  A request can definitely take longer than 20 seconds. 60 seconds of audio might take up 185 seconds to process worst case, though the common case is much much faster.  We should set the timeouts for the worst case.  Same question about all of the timeout_millis below.

File: google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json (hunk 1 above)
Comment (blowmage) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: All these files are all generated by an internal Google tool called GAPIC. The values in this specific file come from the file `google/cloud/speech/v1/cloud_speech_gapic.yaml` in the [googleapis/googleapis](https://github.com/googleapis/googleapis) repo ([link](https://github.com/googleapis/googleapis/blob/master/google/cloud/speech/v1/cloud_speech_gapic.yaml).

I can't point you to documentation on how GAPIC works or why these values were used. Perhaps @swcloud can help here?

File: google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json (hunk 2 above)
Comment (blowmage) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: This file was generated some time ago. The values in the [current configuration](https://github.com/googleapis/googleapis/blob/master/google/cloud/speech/v1/cloud_speech_gapic.yaml) are different now, so we could get some newly generated GAPIC files with updated values. I wouldn't get too hung up on these values for this PR.

**Comment by blowmage** [comment]
@pbogle wrote:

> There are other changes and improvements as well. Adding a link is probably best rather than trying to duplicate the release notes. Please coordinate via email to figure out how we want to do that linking.

I don't understand what you are saying here. Can you clarify?

File: google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json (hunk 1 above)
Comment (swcloud) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: @pbogle Please take a look at https://github.com/googleapis/googleapis/pull/270
Also please discuss with @shinfan regarding whether we should configure retry differently for speech.

File: google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json (hunk 2 above)
Comment (swcloud) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: @landrito Please update the gapic layer to reflect these updated timeout values in the gapic yaml.

Comment (blowmage) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: @landrito You can either submit a PR to the `speech-v1` branch on my personal repo, or you can email me those files and I'll update them.

Comment (landrito) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: @pbogle What exactly is the worst case for this timeout?

Comment (landrito) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: Done.

**Comment by blowmage** [comment]
The GAPIC files have been regenerated to use the latest configuration values.

File: google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json [hunk 3]
```diff
@@ -1,40 +1,40 @@
...
         }
       },
       "methods": {
-        "SyncRecognize": {
+        "Recognize": {
           "timeout_millis": 60000,
```
Comment (pbogle) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: This rpc timeout need to be at least 185 seconds.  (60 * 3 + 5)

Comment (blowmage) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: @landrito What is the process for updating these configuration values? Can we submit a PR to [googleapis/googleapis](https://github.com/googleapis/googleapis)?

Comment (landrito) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: Yep. Submitting a PR to [googleapis/googleapis](https://github.com/googleapis/googleapis) then getting the regenerated gapic.

Comment (blowmage) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: @pbogle Are you going to submit a PR to change this value? Or should we do that? Thoughts, @landrito?

Comment (landrito) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: I'll take care of it (Changing the value -> regen gapic -> delivery)

Comment (landrito) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: Some clarification.

@pbogle are you suggesting the `timeout_millis` value for the methods should be 185 seconds? The rpc timeout milis are already at 190 seconds.

If so, @lukesneeringer  has some context to the `timeout_milis` value for the method.

Comment (lukesneeringer) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: I had edited that value to 190 seconds a week or so ago; it is possible the Ruby GAPIC was generated before that.

**Comment by shinfan** [comment]
@landrito @blowmage I have increased the timeout_mills to 190000 in gapic config: https://github.com/googleapis/googleapis/pull/304

File: google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json (hunk 3 above)
Comment (blowmage) [inline google-cloud-speech/lib/google/cloud/speech/v1/speech_client_config.json]: The GAPIC files were generated last Friday. The values for `initial_rpc_timeout_millis` and `max_rpc_timeout_millis` are 190 seconds, but Recognize's `timeout_millis` is still 60 seconds.

**Comment by blowmage** [comment]
The GAPIC files have been updated again, and the Recognize config is now 190 seconds.

**Comment by blowmage** [comment]
@pbogle The OggOpus and Speex encodings have been added to the documentation. Are there any other changes you want us to make?

**Comment by pbogle** [comment]
In general, does the documentation include all changes to comments in the
v1 proto file?

**Comment by landrito** [comment]
@pbogle, all of the documentation in the proto are reflected in the grpc type documentation under the lib/google/cloud/speech/v1/doc path. 

I did some hand validation, the necessary updates from the proto to the handwritten interface are included as well.

**Comment by pbogle** [comment]
LGTM

**Pull request merged**